### PR TITLE
Feature/pyright drivers

### LINKED ISSUE #578: Resolve Pyright errors for Prompt Drivers

### LINKED ISSUE #579: Resolve Pyright errors for Prompt Model Drivers

### LINKED ISSUE #580: Resolve Pyright errors for SQL Drivers

### LINKED ISSUE #581: Resolve Pyright errors for Vector Drivers

### PR DESCRIPTION
**Pull request by aodhanroche**
closes #578
closes #579
closes #580
closes #581

### CODE REVIEW

File: griptape/drivers/vector/pinecone_vector_store_driver.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def upsert_vector(
 
         params = {"namespace": namespace} | kwargs
 
-        self.index.upsert([(vector_id, vector, meta)], **params)
+        self.index.upsert([(vector_id, vector, meta)], str(params))
```
Comment (cjkindel) [inline griptape/drivers/vector/pinecone_vector_store_driver.py:37]: Is this right? ** generates a dictionary which we are stringifying?

File: griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py [hunk 2]
```diff
@@ -67,14 +67,17 @@ def prompt_stack_to_model_params(self, prompt_stack: PromptStack) -> dict:
             }
         }
 
-    def process_output(self, response_body: str | bytes) -> TextArtifact:
+    def process_output(self, output: list[dict] | str | bytes) -> TextArtifact:
         # When streaming, the response body comes back as bytes.
-        if isinstance(response_body, bytes):
-            response_body = response_body.decode()
+        if isinstance(output, str or bytes):
```
Comment (collindutter) [inline griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py:72]: I'm not sure if this syntax does what we think it does. Let's rewrite to `isinstance(output, str) or isinstance(output, bytes)`

File: griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py [hunk 3]
```diff
@@ -67,14 +67,17 @@ def prompt_stack_to_model_params(self, prompt_stack: PromptStack) -> dict:
...
+            if isinstance(output, bytes):
+                output = output.decode()
 
-        body = json.loads(response_body)
+            body = json.loads(output)
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py:76]: Within `if isinstance(output, bytes):`? I think this location is correct.

File: griptape/drivers/prompt_model/sagemaker_falcon_prompt_model_driver.py [hunk 4]
```diff
@@ -7,17 +8,13 @@
...
-                    "tiiuae/falcon-40b", model_max_length=self.max_tokens
-                )
-            ),
-            takes_self=True,
-        ),
-        kw_only=True,
-    )
+    @property
+    def tokenizer(self) -> HuggingFaceTokenizer:
+        return HuggingFaceTokenizer(
```
Comment (collindutter) [inline griptape/drivers/prompt_model/sagemaker_falcon_prompt_model_driver.py:13]: This will create a new Tokenizer each time the property is accessed, this can be expensive. You should cache the value in a field called `_tokenizer`.

File: griptape/drivers/prompt_model/sagemaker_llama_prompt_model_driver.py [hunk 5]
```diff
@@ -7,17 +8,13 @@
...
-        ),
-        kw_only=True,
-    )
+    @property
+    def tokenizer(self) -> HuggingFaceTokenizer:
+        return HuggingFaceTokenizer(
+            tokenizer=import_optional_dependency("transformers").LlamaTokenizerFast.from_pretrained(
+                "hf-internal-testing/llama-tokenizer", model_max_length=self.max_tokens
+            )
+        )
```
Comment (collindutter) [inline griptape/drivers/prompt_model/sagemaker_llama_prompt_model_driver.py:17]: Same as above, expensive.

File: griptape/drivers/sql/amazon_redshift_sql_driver.py [hunk 6]
```diff
@@ -53,7 +53,7 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
         else:
             return None
 
-    def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]]:
+    def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]] | None:
```
Comment (collindutter) [inline griptape/drivers/sql/amazon_redshift_sql_driver.py:56]: Don't think we want this. And if we did, we'd want to use `Optional`.

File: griptape/drivers/sql/amazon_redshift_sql_driver.py [hunk 7]
```diff
@@ -101,4 +101,4 @@ def get_table_schema(self, table: str, schema: Optional[str] = None) -> Optional
         if self.database_credentials_secret_arn:
             function_kwargs["SecretArn"] = self.database_credentials_secret_arn
         response = self.client.describe_table(**function_kwargs)
-        return [col["name"] for col in response["ColumnList"]]
+        return str([col["name"] for col in response["ColumnList"]])
```
Comment (collindutter) [inline griptape/drivers/sql/amazon_redshift_sql_driver.py:104]: This changes the functionality, previously it was a list, now it's a string.

File: griptape/drivers/sql/snowflake_sql_driver.py [hunk 8]
```diff
@@ -47,16 +47,19 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
...
-            if results.returns_rows:
-                return [{column: value for column, value in result.items()} for result in results]
+            if results is not None:
+                if results.returns_rows:
+                    return [{column: value for column, value in result.items()} for result in results]
+                else:
+                    return None
             else:
-                return None
+                raise ValueError("Results cannot be 'None'")
```
Comment (collindutter) [inline griptape/drivers/sql/snowflake_sql_driver.py:62]: Let's update error with something like `No results found.`

File: griptape/drivers/sql/snowflake_sql_driver.py [hunk 9]
```diff
@@ -65,6 +68,6 @@ def get_table_schema(self, table: str, schema: Optional[str] = None) -> Optional
             metadata_obj = sqlalchemy.MetaData()
             metadata_obj.reflect(bind=self.engine)
             table = sqlalchemy.Table(table, metadata_obj, schema=schema, autoload=True, autoload_with=self.engine)
-            return str([(c.name, c.type) for c in table.columns])
+            return str([(c.name, c.type) for c in getattr(table, "columns")])
```
Comment (collindutter) [inline griptape/drivers/sql/snowflake_sql_driver.py:71]: What's the error that prompted this change?

File: griptape/drivers/sql/sql_driver.py [hunk 10]
```diff
@@ -28,16 +28,19 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
...
-            if results.returns_rows:
-                return [{column: value for column, value in result.items()} for result in results]
+            if results is not None:
+                if results.returns_rows:
+                    return [{column: value for column, value in result.items()} for result in results]
+                else:
+                    return None
             else:
-                return None
+                raise ValueError("results cannot be 'None'")
```
Comment (collindutter) [inline griptape/drivers/sql/sql_driver.py:43]: Same comment as above.

File: griptape/drivers/sql/sql_driver.py [hunk 11]
```diff
@@ -46,6 +49,6 @@ def get_table_schema(self, table: str, schema: Optional[str] = None) -> Optional
             table = sqlalchemy.Table(
                 table, sqlalchemy.MetaData(bind=self.engine), schema=schema, autoload=True, autoload_with=self.engine
             )
-            return str([(c.name, c.type) for c in table.columns])
+            return str([(c.name, c.type) for c in getattr(table, "columns")])
```
Comment (collindutter) [inline griptape/drivers/sql/sql_driver.py:52]: Same comment as above?

File: griptape/drivers/vector/base_vector_store_driver.py [hunk 12]
```diff
@@ -56,7 +56,10 @@ def upsert_text_artifact(
         else:
             vector = artifact.generate_embedding(self.embedding_driver)
 
-        return self.upsert_vector(vector, vector_id=artifact.id, namespace=namespace, meta=meta, **kwargs)
+        if isinstance(vector, list):
+            return self.upsert_vector(vector, vector_id=artifact.id, namespace=namespace, meta=meta, **kwargs)
+        else:
+            raise ValueError("Vector must be of instance 'list'." "")
```
Comment (collindutter) [inline griptape/drivers/vector/base_vector_store_driver.py:62]: Extra `""`?

File: griptape/drivers/vector/marqo_vector_store_driver.py [hunk 13]
```diff
@@ -119,6 +125,7 @@ def load_entries(self, namespace: Optional[str] = None) -> list[BaseVectorStoreD
         """
 
         filter_string = f"namespace:{namespace}" if namespace else None
+
```
Comment (collindutter) [inline griptape/drivers/vector/marqo_vector_store_driver.py:128]: Remove extra line

File: griptape/drivers/vector/opensearch_vector_store_driver.py [hunk 14]
```diff
@@ -154,7 +154,7 @@ def query(
                 id=hit["_id"],
                 namespace=hit["_source"].get("namespace") if namespace else None,
                 score=hit["_score"],
-                vector=hit["_source"].get("vector") if include_vectors else None,
+                vector=hit["_source"]["vector"] if include_vectors else [],
```
Comment (collindutter) [inline griptape/drivers/vector/opensearch_vector_store_driver.py:157]: Let's instead make `vector` optional in `BaseVectorStoreDriver.QueryResult`. Then we can revert this change.

File: griptape/drivers/vector/pgvector_vector_store_driver.py [hunk 15]
```diff
@@ -52,11 +53,11 @@ def validate_engine(self, _, engine: Optional[Engine]) -> None:
             raise ValueError("An engine or connection string is required")
 
     def __attrs_post_init__(self) -> None:
-        """If a an engine is provided, it will be used to connect to the database.
+        """If an engine is provided, it will be used to connect to the database.
         If not, a connection string is used to create a new database connection here.
         """
         if self.engine is None:
-            self.engine = create_engine(self.connection_string, **self.create_engine_params)
+            self.engine = create_engine(self.connection_string, **self.create_engine_params)  # pyright: ignore
```
Comment (collindutter) [inline griptape/drivers/vector/pgvector_vector_store_driver.py:60]: What was the error that prompted the ignore?

File: griptape/drivers/vector/pgvector_vector_store_driver.py [hunk 16]
```diff
@@ -86,15 +87,18 @@ def upsert_vector(
             obj = session.merge(obj)
             session.commit()
 
-            return str(obj.id)
+            return str(getattr(obj, "id"))
```
Comment (collindutter) [inline griptape/drivers/vector/pgvector_vector_store_driver.py:90]: What was the error that prompted this change?

File: griptape/drivers/vector/pgvector_vector_store_driver.py [hunk 17]
```diff
@@ -86,15 +87,18 @@ def upsert_vector(
...
         """Retrieves a specific vector entry from the collection based on its identifier and optional namespace."""
         with Session(self.engine) as session:
             result = session.get(self._model, vector_id)
 
             return BaseVectorStoreDriver.Entry(
-                id=result.id, vector=result.vector, namespace=result.namespace, meta=result.meta
+                id=getattr(result, "id"),
+                vector=getattr(result, "vector"),
+                namespace=getattr(result, "namespace"),
+                meta=getattr(result, "meta"),
```
Comment (collindutter) [inline griptape/drivers/vector/pgvector_vector_store_driver.py:101]: What was the error that prompted these changes?

File: griptape/drivers/vector/pgvector_vector_store_driver.py [hunk 18]
```diff
@@ -142,17 +146,17 @@ def query(
             vector = self.embedding_driver.embed_string(query)
 
             # The query should return both the vector and the distance metric score.
-            query = session.query(self._model, op(vector).label("score")).order_by(op(vector))
+            query_result = session.query(self._model, op(vector).label("score")).order_by(op(vector))  # pyright: ignore
```
Comment (collindutter) [inline griptape/drivers/vector/pgvector_vector_store_driver.py:149]: Is the `pyright: ignore` needed?

File: griptape/drivers/vector/pgvector_vector_store_driver.py [hunk 19]
```diff
@@ -142,17 +146,17 @@ def query(
...
+                query_result = query_result.filter_by(namespace=namespace)
 
-            results = query.limit(count).all()
+            results = query_result.limit(count).all()
 
             return [
                 BaseVectorStoreDriver.QueryResult(
                     id=str(result[0].id),
-                    vector=result[0].vector if include_vectors else None,
+                    vector=result[0].vector if include_vectors else [],
```
Comment (collindutter) [inline griptape/drivers/vector/pgvector_vector_store_driver.py:159]: Shouldn't be needed if we make `vector` Optional as mentioned above.

File: griptape/drivers/vector/pinecone_vector_store_driver.py (hunk 1 above)
Comment (collindutter) [reply griptape/drivers/vector/pinecone_vector_store_driver.py:37]: Yeah this change should be reverted. Curious what the pyright error was?

File: griptape/drivers/vector/redis_vector_store_driver.py [hunk 20]
```diff
@@ -61,7 +60,9 @@ def upsert_vector(
         key = self._generate_key(vector_id, namespace)
         bytes_vector = json.dumps(vector).encode("utf-8")
 
-        mapping = {"vector": np.array(vector, dtype=np.float32).tobytes(), "vec_string": bytes_vector}
+        mapping = {}
+        mapping["vector"] = np.array(vector, dtype=np.float32).tobytes()
+        mapping["vec_string"] = bytes_vector
```
Comment (collindutter) [inline griptape/drivers/vector/redis_vector_store_driver.py:65]: Why this change?

File: griptape/drivers/vector/redis_vector_store_driver.py [hunk 21]
```diff
@@ -129,14 +130,14 @@ def query(
 
         query_params = {"vector": np.array(vector, dtype=np.float32).tobytes()}
 
-        results = self.client.ft(self.index).search(query_expression, query_params).docs
+        results = self.client.ft(self.index).search(query_expression, query_params).docs  # pyright: ignore
```
Comment (collindutter) [inline griptape/drivers/vector/redis_vector_store_driver.py:133]: Why the pyright ignore?

File: griptape/drivers/vector/redis_vector_store_driver.py [hunk 22]
```diff
@@ -129,14 +130,14 @@ def query(
...
 
         query_results = []
         for document in results:
             metadata = getattr(document, "metadata", None)
             namespace = document.id.split(":")[0] if ":" in document.id else None
             vector_id = document.id.split(":")[1] if ":" in document.id else document.id
-            vector_float_list = json.loads(document["vec_string"]) if include_vectors else None
+            vector_float_list = json.loads(document["vec_string"]) if include_vectors else []
```
Comment (collindutter) [inline griptape/drivers/vector/redis_vector_store_driver.py:140]: Should be able to revert with the optional vector change.

Comment (collindutter) [inline griptape/drivers/vector/redis_vector_store_driver.py:140]: Should be able to revert with the optional vector change.

File: tests/unit/drivers/sql/test_amazon_redshift_sql_driver.py [hunk 23]
```diff
@@ -162,4 +162,4 @@ def test_execute_query(self, statement_driver):
         assert statement_driver.execute_query("query") == [BaseSqlDriver.RowResult(row) for row in rows]
 
     def test_get_table_schema(self, describe_table_driver):
-        assert describe_table_driver.get_table_schema("dev") == ["first_name", "last_name"]
+        assert describe_table_driver.get_table_schema("dev") == "['first_name', 'last_name']"
```
Comment (collindutter) [inline tests/unit/drivers/sql/test_amazon_redshift_sql_driver.py:165]: This doesn't feel quite right, `get_table_schema` should still return a list.

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjkindel) [reply griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py:76]: Tabbing just looked off within the diff, reading just the new code makes sense

File: griptape/drivers/sql/snowflake_sql_driver.py (hunk 9 above)
Comment (aodhanroche) [reply griptape/drivers/sql/snowflake_sql_driver.py:71]: Cannot access member "columns" for type "str"

File: griptape/drivers/sql/amazon_redshift_sql_driver.py (hunk 6 above)
Comment (aodhanroche) [reply griptape/drivers/sql/amazon_redshift_sql_driver.py:56]: `Optional` also solve the two errors that this changed was fixing. Those are 
`amazon_redshift_sql_driver.py:56:48 - error: Function with declared return type "list[dict[str, Any | None]]" must return value on all code paths "None" is incompatible with "list[dict[str, Any | None]]" (reportReturnType)`
  and
  `amazon_redshift_sql_driver.py:89:20 - error: Expression of type "None" cannot be assigned to return type "list[dict[str, Any | None]]"
    "None" is incompatible with "list[dict[str, Any | None]]" (reportReturnType)`
 for reference

File: griptape/drivers/sql/amazon_redshift_sql_driver.py (hunk 7 above)
Comment (aodhanroche) [reply griptape/drivers/sql/amazon_redshift_sql_driver.py:104]: Should I change the return type of `get_table_schema()` then? Its currently returning `Optional[str]` which is the problem

File: griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py [hunk 24]
```diff
@@ -67,14 +67,17 @@ def prompt_stack_to_model_params(self, prompt_stack: PromptStack) -> dict:
...
 
-        if self.prompt_driver.stream:
-            return TextArtifact(body["outputText"])
+            if self.prompt_driver.stream:
+                return TextArtifact(body["outputText"])
+            else:
+                return TextArtifact(body["results"][0]["outputText"])
         else:
-            return TextArtifact(body["results"][0]["outputText"])
+            raise ValueError("output must be of instance 'str' or 'bytes'")
```
[the remark this answers is not in the document]
Comment (aodhanroche) [reply griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py]: Sounds good to me, will change all my error messages like this to `of type`

File: griptape/drivers/vector/pinecone_vector_store_driver.py (hunk 1 above)
Comment (aodhanroche) [reply griptape/drivers/vector/pinecone_vector_store_driver.py:37]: This was 2 pyright errors:
`Argument of type "str | None" cannot be assigned to parameter "batch_size" of type "int | None" in function "upsert"` and `Argument of type "str | None" cannot be assigned to parameter "show_progress" of type "bool" in function "upsert"`

File: griptape/drivers/sql/sql_driver.py (hunk 11 above)
Comment (aodhanroche) [reply griptape/drivers/sql/sql_driver.py:52]: same error as above

File: griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/prompt_model/bedrock_titan_prompt_model_driver.py]: `of type` has a slightly different meaning. Let's do `an instance of`.

File: griptape/drivers/vector/base_vector_store_driver.py [hunk 25]
```diff
@@ -56,15 +56,19 @@ def upsert_text_artifact(
...
+                             "")
 
     def upsert_text(
-        self,
-        string: str,
-        vector_id: Optional[str] = None,
-        namespace: Optional[str] = None,
-        meta: Optional[dict] = None,
-        **kwargs,
+            self,
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/vector/base_vector_store_driver.py]: I've got no clue...This doesn't seem right, but the code check CI step isn't failing..

File: tests/unit/drivers/sql/test_amazon_redshift_sql_driver.py (hunk 23 above)
Comment (aodhanroche) [reply tests/unit/drivers/sql/test_amazon_redshift_sql_driver.py:165]: Change was made to match the changes in `amazon_redshift_sql_driver`

File: griptape/drivers/vector/pgvector_vector_store_driver.py (hunk 15 above)
Comment (aodhanroche) [reply griptape/drivers/vector/pgvector_vector_store_driver.py:60]: `Cannot assign member "engine" for type "PgVectorVectorStoreDriver*" ` which seems to point to a problem with SQL Alchemy

File: griptape/drivers/vector/pgvector_vector_store_driver.py (hunk 16 above)
Comment (aodhanroche) [reply griptape/drivers/vector/pgvector_vector_store_driver.py:90]: `Cannot access member "id" for type "_symbol"`

File: griptape/drivers/vector/pgvector_vector_store_driver.py (hunk 17 above)
Comment (aodhanroche) [reply griptape/drivers/vector/pgvector_vector_store_driver.py:101]: Same as above (`Cannot access member "____" for type "_symbol"`) on each of the attributes

File: griptape/drivers/vector/pgvector_vector_store_driver.py (hunk 18 above)
Comment (aodhanroche) [reply griptape/drivers/vector/pgvector_vector_store_driver.py:149]: Similar issue to the one I'm assuming is on the SQL Alchemy end... `Object of type "None" cannot be called` which I think is pointing at `self.engine` possibly being `None`

File: griptape/drivers/vector/redis_vector_store_driver.py (hunk 21 above)
Comment (aodhanroche) [reply griptape/drivers/vector/redis_vector_store_driver.py:133]: `Argument of type "dict[str, bytes]" cannot be assigned to parameter "query_params" of type "_QueryParams | None" in function "search"` was the error, ignored hoping for some insight here

File: griptape/drivers/vector/redis_vector_store_driver.py (hunk 20 above)
Comment (aodhanroche) [reply griptape/drivers/vector/redis_vector_store_driver.py:65]: @andrewfrench helped me out with this one, but it solves both `Argument of type "str" cannot be assigned to parameter "__value" of type "bytes" in function "__setitem__"` and `Argument of type "dict[str, bytes]" cannot be assigned to parameter "mapping" of type "Mapping[_Key, _Value]" in function "hset"`

Comment (andrewfrench) [reply griptape/drivers/vector/redis_vector_store_driver.py:65]: In all honesty this fix was more the result of trial and error than anything. The Pyright complaint was pretty resistant to any change except defining each key-value pair independently.

File: griptape/drivers/prompt_model/sagemaker_falcon_prompt_model_driver.py [hunk 26]
```diff
@@ -7,17 +8,17 @@
...
 @define
 class SageMakerFalconPromptModelDriver(BasePromptModelDriver):
-    tokenizer: BaseTokenizer = field(
-        default=Factory(
-            lambda self: HuggingFaceTokenizer(
+    _tokenizer: HuggingFaceTokenizer = field(default=None, kw_only=True)
+
+    @property
+    def tokenizer(self) -> HuggingFaceTokenizer:
+        if not self._tokenizer:
```
Comment (collindutter) [inline griptape/drivers/prompt_model/sagemaker_falcon_prompt_model_driver.py:15]: Better to be explicit when checking for `None`. Replace with `if self._tokenizer is None`.

File: griptape/drivers/prompt_model/sagemaker_llama_prompt_model_driver.py [hunk 27]
```diff
@@ -7,17 +8,17 @@
...
 @define
 class SageMakerLlamaPromptModelDriver(BasePromptModelDriver):
-    tokenizer: BaseTokenizer = field(
-        default=Factory(
-            lambda self: HuggingFaceTokenizer(
+    _tokenizer: HuggingFaceTokenizer = field(default=None, kw_only=True)
+
+    @property
+    def tokenizer(self) -> HuggingFaceTokenizer:
+        if not self._tokenizer:
```
Comment (collindutter) [inline griptape/drivers/prompt_model/sagemaker_llama_prompt_model_driver.py:15]: Same comment.

File: griptape/drivers/sql/amazon_redshift_sql_driver.py (hunk 7 above)
Comment (collindutter) [reply griptape/drivers/sql/amazon_redshift_sql_driver.py:104]: Oh great point, ok let's continue with this then.

File: griptape/drivers/sql/snowflake_sql_driver.py [hunk 28]
```diff
@@ -47,16 +47,19 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
         else:
             return None
 
-    def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]]:
+    def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]] | None:
```
Comment (collindutter) [inline griptape/drivers/sql/snowflake_sql_driver.py:50]: Let's revert this change with my change below.

File: griptape/drivers/sql/snowflake_sql_driver.py [hunk 29]
```diff
@@ -47,16 +47,19 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
...
         with self.engine.connect() as con:
             results = con.execute(sqlalchemy.text(query))
 
-            if results.returns_rows:
-                return [{column: value for column, value in result.items()} for result in results]
+            if results is not None:
+                if results.returns_rows:
+                    return [{column: value for column, value in result.items()} for result in results]
+                else:
+                    return None
```
Comment (collindutter) [inline griptape/drivers/sql/snowflake_sql_driver.py:60]: Let's raise an error here as well.

File: griptape/drivers/sql/snowflake_sql_driver.py (hunk 9 above)
Comment (collindutter) [reply griptape/drivers/sql/snowflake_sql_driver.py:71]: Can you try this? https://stackoverflow.com/questions/24959589/get-table-columns-from-sqlalchemy-table-model

File: griptape/drivers/sql/sql_driver.py [hunk 30]
```diff
@@ -28,16 +28,19 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
...
-            if results.returns_rows:
-                return [{column: value for column, value in result.items()} for result in results]
+            if results is not None:
+                if results.returns_rows:
+                    return [{column: value for column, value in result.items()} for result in results]
+                else:
+                    return None
             else:
-                return None
+                raise ValueError("No result found")
```
Comment (collindutter) [inline griptape/drivers/sql/sql_driver.py:43]: Same point as above.

File: griptape/drivers/sql/sql_driver.py (hunk 11 above)
Comment (collindutter) [reply griptape/drivers/sql/sql_driver.py:52]: Let's try what I suggested in the other comment.

File: griptape/drivers/vector/pgvector_vector_store_driver.py (hunk 15 above)
Comment (collindutter) [reply griptape/drivers/vector/pgvector_vector_store_driver.py:60]: Let's try this:
`from typing import cast`

`self.engine = cast(Engine, create_engine(self.connection_string, **self.create_engine_params))`

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/drivers/sql/snowflake_sql_driver.py [hunk 31]
```diff
@@ -53,18 +53,21 @@ def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]]:
...
     def get_table_schema(self, table: str, schema: Optional[str] = None) -> Optional[str]:
         sqlalchemy = import_optional_dependency("sqlalchemy")
 
         try:
             metadata_obj = sqlalchemy.MetaData()
             metadata_obj.reflect(bind=self.engine)
-            table = sqlalchemy.Table(table, metadata_obj, schema=schema, autoload=True, autoload_with=self.engine)
-            return str([(c.name, c.type) for c in table.columns])
+            table_name = sqlalchemy.Table(table, metadata_obj, schema=schema, autoload=True, autoload_with=self.engine)
+            return str([(c.name, c.type) for c in table_name.columns])
```
Comment (collindutter) [inline griptape/drivers/sql/snowflake_sql_driver.py:71]: Revert to `table`. I meant to try `table.__table__.columns`

Comment (aodhanroche) [reply griptape/drivers/sql/snowflake_sql_driver.py]: Doesn't solve, #ignoring

File: griptape/drivers/sql/sql_driver.py [hunk 32]
```diff
@@ -28,16 +28,19 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
...
         with self.engine.begin() as con:
             results = con.execute(sqlalchemy.text(query))
 
-            if results.returns_rows:
-                return [{column: value for column, value in result.items()} for result in results]
+            if results is not None:
+                if results.returns_rows:
+                    return [{column: value for column, value in result.items()} for result in results]
+                else:
+                    return None
```
Comment (collindutter) [inline griptape/drivers/sql/sql_driver.py:41]: Raise an error.

File: griptape/drivers/sql/sql_driver.py [hunk 33]
```diff
@@ -28,16 +28,19 @@ def execute_query(self, query: str) -> Optional[list[BaseSqlDriver.RowResult]]:
         else:
             return None
 
-    def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]]:
+    def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]] | None:
```
Comment (collindutter) [inline griptape/drivers/sql/sql_driver.py:31]: Remove `| None`

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/drivers/sql/snowflake_sql_driver.py [hunk 34]
```diff
@@ -53,10 +53,13 @@ def execute_query_raw(self, query: str) -> list[dict[str, Optional[Any]]]:
         with self.engine.connect() as con:
             results = con.execute(sqlalchemy.text(query))
 
-            if results.returns_rows:
-                return [{column: value for column, value in result.items()} for result in results]
+            if results is not None:
+                if results.returns_rows:
+                    return [{column: value for column, value in result.items()} for result in results]
+                else:
+                    raise ValueError("No results found")
```
Comment (collindutter) [inline griptape/drivers/sql/snowflake_sql_driver.py:60]: This is now inconsistent with the parent class.

File: griptape/drivers/vector/marqo_vector_store_driver.py [hunk 35]
```diff
@@ -204,7 +214,7 @@ def get_indexes(self) -> list[str]:
         """
 
         # Change this once API issue is fixed (entries in results are no longer objects but dicts)
```
Comment (collindutter) [inline griptape/drivers/vector/marqo_vector_store_driver.py:216]: Remove comment

File: tests/unit/drivers/vector/test_marqo_vector_store_driver.py [hunk 36]
```diff
@@ -193,7 +193,7 @@ def test_load_entries(self, driver, mock_marqo):
 
         # Assert
         assert len(entries) == 1
-        mock_marqo.index().search.assert_called_once_with("", limit=10000, filter_string=None)
+        mock_marqo.index().search.assert_called_once_with("", limit=10000)
```
Comment (collindutter) [inline tests/unit/drivers/vector/test_marqo_vector_store_driver.py:196]: Add a test that specifies `namespace` so that the `filter_string` case is still handled.

**Review by collindutter — CHANGES_REQUESTED** [review]

**Review by andrewfrench — CHANGES_REQUESTED** [review]
Hats off to your work in the trenches of the Python typing wars. One small oversight, probably from our pairing session.

File: griptape/drivers/vector/opensearch_vector_store_driver.py [hunk 37]
```diff
@@ -154,7 +154,7 @@ def query(
                 id=hit["_id"],
                 namespace=hit["_source"].get("namespace") if namespace else None,
                 score=hit["_score"],
-                vector=hit["_source"].get("vector") if include_vectors else None,
+                vector=hit["_source"].get["vector"] if include_vectors else None,
```
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/vector/opensearch_vector_store_driver.py]: @aodhanroche this should be `hit["_source"].get("vector")`

**Review by collindutter — APPROVED** [review]
